{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyzing genotype-phenotype maps\n",
    "Example and homework for Genome Sciences 541, by Jesse Bloom, June-2019.\n",
    "\n",
    "The questions in this notebook are due by **5 PM on Thursday June 13**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.ion()\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import scipy.optimize"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create an additive genotype-phenotype map\n",
    "Define a class that sets up an additive model for phenotypes and randomly assigns effects to each mutation at each site:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class AdditiveModel:\n",
    "    \"\"\"Additive genotype-phenotype model.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    alphabet : list\n",
    "        List of characters in the alphabet for the sequences.\n",
    "    wtseq : str\n",
    "        The wildtype sequence.\n",
    "    seed : int\n",
    "        The random number seed for initializing model parameters\n",
    "    \n",
    "    Attributes\n",
    "    ----------\n",
    "    alphabet : list\n",
    "        List of characters in the alphabet for the sequences.\n",
    "    wtseq : str\n",
    "        The wildtype sequence.\n",
    "    seqlength : int\n",
    "        Length of sequences.\n",
    "    wtpheno : float\n",
    "        Phenotype of wildtype sequence.\n",
    "    site_effects : dict\n",
    "        Keyed by `(site, a)` where `site` is site in 0, 1, ... numbering and\n",
    "        `s` is each possible letter in alphabet; values are effects of mutations.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self,\n",
    "                 alphabet,\n",
    "                 wtseq,\n",
    "                 seed=1):\n",
    "        \"\"\"See main class docstring for how to initialize.\"\"\"\n",
    "        self.alphabet = alphabet\n",
    "        self.wtseq = wtseq\n",
    "        for a in wtseq:\n",
    "            if a not in alphabet:\n",
    "                raise ValueError(f\"wtseq contains character not in alphabet: {a}\")\n",
    "        self.seqlength = len(wtseq)\n",
    "        \n",
    "        scipy.random.seed(seed)\n",
    "        self.wtpheno = scipy.random.normal(2)\n",
    "        self.site_effects = {}\n",
    "        for r, a_wt in enumerate(self.wtseq):\n",
    "            for a in self.alphabet:\n",
    "                if a == a_wt:\n",
    "                    self.site_effects[(r, a)] = 0\n",
    "                else:\n",
    "                    self.site_effects[(r, a)] = scipy.random.normal(-1, 2)\n",
    "                    \n",
    "    def muteffects_table(self):\n",
    "        \"\"\"Table of effects of all mutations\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        pandas DataFrame\n",
    "            Columns give site, wildtype identity, mutation, and its effect.\n",
    "        \"\"\"\n",
    "        return (pd.Series(self.site_effects)\n",
    "                .rename_axis(['site', 'mutation'])\n",
    "                .reset_index(name='muteffect')\n",
    "                .assign(wildtype=lambda x: x['site'].apply(lambda r: self.wtseq[r]))\n",
    "                [['site', 'wildtype', 'mutation', 'muteffect']]\n",
    "                .query('wildtype != mutation')\n",
    "                .reset_index(drop=True)\n",
    "                )\n",
    "    \n",
    "    def pheno(self, seq):\n",
    "        \"\"\"Returns phenotype of sequence.\n",
    "        \n",
    "        Parameters\n",
    "        ------------\n",
    "        seq : str\n",
    "            A sequence.\n",
    "            \n",
    "        Returns\n",
    "        ---------\n",
    "        float\n",
    "            The phenotype.\n",
    "        \"\"\"\n",
    "        return sum(self.site_effects[(r, a)] for r, a in enumerate(seq)) + self.wtpheno\n",
    "    \n",
    "    def all_seqs_phenos(self):\n",
    "        \"\"\"Get all possible sequences and their phenotypes.\n",
    "        \n",
    "        Returns\n",
    "        ---------\n",
    "        dict\n",
    "            Keyed by all possible sequences, values are their phenotypes.\n",
    "        \"\"\"\n",
    "        # recursive function to build all sequences\n",
    "        def _add_letter(seqlist):\n",
    "            if len(seqlist[0]) == self.seqlength:\n",
    "                return seqlist\n",
    "            newseqlist = []\n",
    "            for seq in seqlist:\n",
    "                for a in self.alphabet:\n",
    "                    newseqlist.append(seq + a)\n",
    "            return _add_letter(newseqlist)\n",
    "            \n",
    "        allseqs = _add_letter(self.alphabet)\n",
    "        return {seq: self.pheno(seq) for seq in allseqs}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now create an additive model of binary sequences of letters `0` and `1` where the wildtype sequence is `0000`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "binary_additive_model = AdditiveModel(['0', '1'], '0000')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are the effects of mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>muteffect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.223513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.056344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-3.145937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.730815</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site wildtype mutation  muteffect\n",
       "0     0        0        1  -2.223513\n",
       "1     1        0        1  -2.056344\n",
       "2     2        0        1  -3.145937\n",
       "3     3        0        1   0.730815"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binary_additive_model.muteffects_table()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are all the genotypes and their phenotypes in a dict:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0000': 3.6243453636632417,\n",
       " '0001': 4.3551606223125985,\n",
       " '0010': 0.47840811935090066,\n",
       " '0011': 1.2092233780002575,\n",
       " '0100': 1.5680018591363303,\n",
       " '0101': 2.2988171177856875,\n",
       " '0110': -1.5779353851760103,\n",
       " '0111': -0.8471201265266535,\n",
       " '1000': 1.4008325363630911,\n",
       " '1001': 2.131647795012448,\n",
       " '1010': -1.7451047079492499,\n",
       " '1011': -1.014289449299893,\n",
       " '1100': -0.6555109681638207,\n",
       " '1101': 0.07530429048553611,\n",
       " '1110': -3.8014482124761617,\n",
       " '1111': -3.070632953826805}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binary_additive_gp = binary_additive_model.all_seqs_phenos()\n",
    "binary_additive_gp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is the same in a nice data frame format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0000</td>\n",
       "      <td>3.624345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0001</td>\n",
       "      <td>4.355161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0010</td>\n",
       "      <td>0.478408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0011</td>\n",
       "      <td>1.209223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0100</td>\n",
       "      <td>1.568002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0101</td>\n",
       "      <td>2.298817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0110</td>\n",
       "      <td>-1.577935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0111</td>\n",
       "      <td>-0.847120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1000</td>\n",
       "      <td>1.400833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1001</td>\n",
       "      <td>2.131648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1010</td>\n",
       "      <td>-1.745105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1011</td>\n",
       "      <td>-1.014289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1100</td>\n",
       "      <td>-0.655511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1101</td>\n",
       "      <td>0.075304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1110</td>\n",
       "      <td>-3.801448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1111</td>\n",
       "      <td>-3.070633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genotype  phenotype\n",
       "0      0000   3.624345\n",
       "1      0001   4.355161\n",
       "2      0010   0.478408\n",
       "3      0011   1.209223\n",
       "4      0100   1.568002\n",
       "5      0101   2.298817\n",
       "6      0110  -1.577935\n",
       "7      0111  -0.847120\n",
       "8      1000   1.400833\n",
       "9      1001   2.131648\n",
       "10     1010  -1.745105\n",
       "11     1011  -1.014289\n",
       "12     1100  -0.655511\n",
       "13     1101   0.075304\n",
       "14     1110  -3.801448\n",
       "15     1111  -3.070633"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binary_additive_gp_df = (\n",
    "    pd.Series(binary_additive_gp)\n",
    "    .rename_axis('genotype')\n",
    "    .reset_index(name='phenotype')\n",
    "    )\n",
    "binary_additive_gp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is the distribution of phenotypes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f564ce11a90>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAD4CAYAAADhNOGaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAARY0lEQVR4nO3de4yldX3H8feHZSt4g9QdK90Lo5XYWi+IK2KMKdXarKhQrbYY77VuYzVqalMBDV4aE42tWEXFrRBRqTdAXQWjEImXpAILcl+tG7WySAuCsqAUuvjtH+cZMx3O7p5d9znP7Pzer+TE5/Kbcz47MvOZ5/k95zmpKiRJ7dpv6ACSpGFZBJLUOItAkhpnEUhS4ywCSWrc/kMH2F0rVqyo2dnZoWNI0j7lsssu+2lVzYzbt88VwezsLJs2bRo6hiTtU5L85472eWpIkhpnEUhS4ywCSWqcRSBJjbMIJKlxFoEkNa63IkhyQJJLklyZ5Nokbxsz5j5JPp1kS5KLk8z2lUeSNF6fRwR3AU+tqscChwPrkhy1YMwrgJ9V1cOBU4B39ZhHkjRGb0VQI3d0q8u7x8IPPzgOOLNbPht4WpL0lUmSdG+9vrM4yTLgMuDhwAeq6uIFQ1YC1wNU1fYktwEPAn664HnWA+sB1qxZ02dk7UWzJ5w3yOv+6J3PHOR1h+T3Wr+JXieLq+qeqjocWAUcmeRRe/g8G6pqbVWtnZkZe6sMSdIemspVQ1X1c+AiYN2CXTcAqwGS7A8cBNwyjUySpJE+rxqaSXJwt3wg8HTguwuGbQRe2i0/D/ha+SHKkjRVfc4RHAKc2c0T7Ad8pqq+lOTtwKaq2gicDnw8yRbgVuD4HvNIksborQiq6irgcWO2nzxv+X+A5/eVQZK0a76zWJIaZxFIUuMsAklqnEUgSY2zCCSpcRaBJDXOIpCkxlkEktQ4i0CSGmcRSFLjLAJJapxFIEmNswgkqXEWgSQ1ziKQpMZZBJLUOItAkhpnEUhS4ywCSWqcRSBJjbMIJKlxFoEkNc4ikKTGWQSS1DiLQJIa11sRJFmd5KIk1yW5Nsnrxow5OsltSa7oHif3lUeSNN7+PT73duANVXV5kgcAlyW5oKquWzDum1X1rB5zSJJ2orcjgqq6saou75ZvBzYDK/t6PUnSnpnKHEGSWeBxwMVjdj8pyZVJvpzkD3fw9euTbEqy6eabb+4xqSS1p/ciSHJ/4Bzg9VW1bcHuy4FDq+qxwPuBz497jqraUFVrq2rtzMxMv4ElqTG9FkGS5YxK4KyqOnfh/qraVlV3dMvnA8uTrOgzkyTp/+vzqqEApwObq+o9OxjzkG4cSY7s8tzSVyZJ0r31edXQk4EXA1cnuaLbdhKwBqCqTgOeB7wqyXbgTuD4qqoeM0mSFuitCKrqW0B2MeZU4NS+MkiSds13FktS4ywCSWqcRSBJjbMIJKlxFoEkNc4ikKTGWQSS1DiLQJIaZxFIUuMsAklqnEUgSY2zCCSpcRaBJDXOIpCkxlkEktQ4i0CSGmcRSFLjLAJJapxFIEmNswgkqXEWgSQ1ziKQpMZZBJLUOItAkhpnEUhS4ywCSWpcb0WQZHWSi5Jcl+TaJK8bMyZJ3pdkS5KrkhzRVx5J0nj79/jc24E3VNXlSR4AXJbkgqq6bt6YZwCHdY8nAh/q/leSNCW9HRFU1Y1VdXm3fDuwGVi5YNhxwMdq5NvAwUkO6SuTJOne+jwi+LUks8DjgIsX7FoJXD9vfWu37cYFX78eWA+wZs2aPc4xe8J5e/y1v6kfvfOZg7zukP9mLX3+TE1XX//m3ieLk9wfOAd4fVVt25PnqKoNVbW2qtbOzMzs3YCS1LheiyDJckYlcFZVnTtmyA3A6nnrq7ptkqQp6fOqoQCnA5ur6j07GLYReEl39dBRwG1VdeMOxkqSetDnHMGTgRcDVye5ott2ErAGoKpOA84HjgG2AL8EXt5jHknSGBMVQZJHV9XVu/PEVfUtILsYU8Crd+d5JUl716Snhj6Y5JIkf5vkoF4TSZKmaqIiqKqnAC9kNLF7WZJ/S/L0XpNJkqZi4sniqvo+8GbgjcAfAe9L8t0kz+0rnCSpfxMVQZLHJDmF0buDnwo8u6r+oFs+pcd8kqSeTXrV0PuBjwAnVdWdcxur6idJ3txLMknSVExaBM8E7qyqewCS7AccUFW/rKqP95ZOktS7SecILgQOnLd+326bJGkfN2kRHFBVd8ytdMv37SeSJGmaJi2CX8z/0Jgkjwfu3Ml4SdI+YtI5gtcDn03yE0bvFn4I8Je9pZIkTc1ERVBVlyb5feAR3abvVdX/9hdLkjQtu3PTuScAs93XHJGEqvpYL6kkSVMz6U3nPg78HnAFcE+3uQCLQJL2cZMeEawFHtndLVSStIRMetXQNYwmiCVJS8ykRwQrgOuSXALcNbexqo7tJZUkaWomLYK39hlCkjScSS8f/XqSQ4HDqurCJPcFlvUbTZI0DZPehvqVwNnAh7tNK4HP9xVKkjQ9k04Wv5rRh9Fvg19/SM2D+wolSZqeSYvgrqq6e24lyf6M3kcgSdrHTVoEX09yEnBg91nFnwW+2F8sSdK0TFoEJwA3A1cDfwOcz+jziyVJ+7hJrxr6FfCv3UOStIRMeq+hHzJmTqCqHrbXE0mSpmp37jU05wDg+cBv7+wLkpwBPAu4qaoeNWb/0cAXgB92m86tqrdPmEeStJdMNEdQVbfMe9xQVe9l9IH2O/NRYN0uxnyzqg7vHpaAJA1g0lNDR8xb3Y/REcJOv7aqvpFkdo+TSZKmYtJTQ/88b3k78CPgL/bC6z8pyZXAT4C/r6prxw1Ksh5YD7BmzZq98LKSpDmTXjX0xz289uXAoVV1R5JjGN2y4rAdvP4GYAPA2rVrfSObJO1Fk54a+rud7a+q9+zuC1fVtnnL5yf5YJIVVfXT3X0uSdKe252rhp4AbOzWnw1cAnx/T184yUOA/66qSnIko7mHW/b0+SRJe2bSIlgFHFFVtwMkeStwXlW9aEdfkOSTwNHAiiRbgbcAywGq6jTgecCrkmwH7gSO96MwJWn6Ji2C3wHunrd+d7dth6rqBbvYfypw6oSvL0nqyaRF8DHgkiSf69b/DDizn0iSpGma9KqhdyT5MvCUbtPLq+o7/cWSJE3LpHcfBbgvsK2q/gXYmuShPWWSJE3RpB9V+RbgjcCJ3ablwCf6CiVJmp5JjwieAxwL/AKgqn4CPKCvUJKk6Zm0CO7uLu0sgCT36y+SJGmaJi2CzyT5MHBwklcCF+KH1EjSkjDpVUP/1H1W8TbgEcDJVXVBr8kkSVOxyyJIsgy4sLvxnL/8JWmJ2eWpoaq6B/hVkoOmkEeSNGWTvrP4DuDqJBfQXTkEUFWv7SWVJGlqJi2Cc7uHJGmJ2WkRJFlTVT+uKu8rJElL1K7mCD4/t5DknJ6zSJIGsKsiyLzlh/UZRJI0jF0VQe1gWZK0ROxqsvixSbYxOjI4sFumW6+qemCv6SRJvdtpEVTVsmkFkSQNY3c+j0CStARZBJLUOItAkhpnEUhS4ywCSWqcRSBJjbMIJKlxFoEkNa63IkhyRpKbklyzg/1J8r4kW5JcleSIvrJIknaszyOCjwLrdrL/GcBh3WM98KEes0iSdqC3IqiqbwC37mTIccDHauTbwMFJDukrjyRpvEk/oawPK4Hr561v7bbduHBgkvWMjhpYs2bNVMJp3zV7wnlDR5D2KfvEZHFVbaiqtVW1dmZmZug4krSkDFkENwCr562v6rZJkqZoyCLYCLyku3roKOC2qrrXaSFJUr96myNI8kngaGBFkq3AW4DlAFV1GnA+cAywBfgl8PK+skiSdqy3IqiqF+xifwGv7uv1JUmT2ScmiyVJ/bEIJKlxFoEkNc4ikKTGWQSS1DiLQJIaZxFIUuMsAklqnEUgSY2zCCSpcRaBJDXOIpCkxlkEktQ4i0CSGmcRSFLjLAJJapxFIEmNswgkqXEWgSQ1ziKQpMZZBJLUOItAkhpnEUhS4ywCSWqcRSBJjeu1CJKsS/K9JFuSnDBm/8uS3Jzkiu7x133mkSTd2/59PXGSZcAHgKcDW4FLk2ysqusWDP10Vb2mrxySpJ3r84jgSGBLVf2gqu4GPgUc1+PrSZL2QJ9FsBK4ft761m7bQn+e5KokZydZPe6JkqxPsinJpptvvrmPrJLUrKEni78IzFbVY4ALgDPHDaqqDVW1tqrWzszMTDWgJC11fRbBDcD8v/BXddt+rapuqaq7utWPAI/vMY8kaYw+i+BS4LAkD03yW8DxwMb5A5IcMm/1WGBzj3kkSWP0dtVQVW1P8hrgK8Ay4IyqujbJ24FNVbUReG2SY4HtwK3Ay/rKI0kar7ciAKiq84HzF2w7ed7yicCJfWaQJO3c0JPFkqSBWQSS1DiLQJIaZxFIUuMsAklqnEUgSY2zCCSpcRaBJDXOIpCkxlkEktQ4i0CSGmcRSFLjLAJJapxFIEmNswgkqXEWgSQ1ziKQpMZZBJLUOItAkhpnEUhS4ywCSWqcRSBJjbMIJKlxFoEkNc4ikKTGWQSS1LheiyDJuiTfS7IlyQlj9t8nyae7/Rcnme0zjyTp3norgiTLgA8AzwAeCbwgySMXDHsF8LOqejhwCvCuvvJIksbr84jgSGBLVf2gqu4GPgUct2DMccCZ3fLZwNOSpMdMkqQF9u/xuVcC189b3wo8cUdjqmp7ktuABwE/nT8oyXpgfbd6R5Lv7eR1Vyz8+kVgRd616DLBIv1eYaZJLcZcU8uUyc8fLMbvE+xBrt34N49z6I529FkEe01VbQA2TDI2yaaqWttzpN2yGDPB4sxlpsktxlxmmtxiytXnqaEbgNXz1ld128aOSbI/cBBwS4+ZJEkL9FkElwKHJXlokt8Cjgc2LhizEXhpt/w84GtVVT1mkiQt0Nupoe6c/2uArwDLgDOq6tokbwc2VdVG4HTg40m2ALcyKovf1ESnkKZsMWaCxZnLTJNbjLnMNLlFkyv+AS5JbfOdxZLUOItAkhq3pIsgyRuSVJIViyDLPya5KskVSb6a5HcXQaZ3J/lul+tzSQ4eOhNAkucnuTbJr5IMenndrm6TMoQkZyS5Kck1Q2cBSLI6yUVJruv+f3vd0JkAkhyQ5JIkV3a53jZ0pjlJliX5TpIvDZ0FlnARJFkN/Cnw46GzdN5dVY+pqsOBLwEnDx0IuAB4VFU9BvgP4MSB88y5Bngu8I0hQ0x4m5QhfBRYN3SIebYDb6iqRwJHAa9eJN+nu4CnVtVjgcOBdUmOGjjTnNcBm4cOMWfJFgGjexf9A7AoZsOratu81fuxCHJV1Veranu3+m1G7/UYXFVtrqqdvXt8Wia5TcrUVdU3GF1ltyhU1Y1VdXm3fDujX3Arh00FNXJHt7q8ewz+c5dkFfBM4CNDZ5mzJIsgyXHADVV15dBZ5kvyjiTXAy9kcRwRzPdXwJeHDrHIjLtNyuC/4Baz7g7CjwMuHjbJSHcK5grgJuCCqloMud7L6I/UXw0dZM4+cYuJcZJcCDxkzK43AScxOi00VTvLVFVfqKo3AW9KciLwGuAtQ2fqxryJ0eH9WX3n2Z1c2rckuT9wDvD6BUfAg6mqe4DDu/mvzyV5VFUNNreS5FnATVV1WZKjh8qx0D5bBFX1J+O2J3k08FDgyu5GpquAy5McWVX/NUSmMc4CzmcKRbCrTEleBjwLeNo039W9G9+rIU1ymxQBSZYzKoGzqurcofMsVFU/T3IRo7mVISfZnwwcm+QY4ADggUk+UVUvGjDT0js1VFVXV9WDq2q2qmYZHc4f0XcJ7EqSw+atHgd8d6gsc5KsY3SIemxV/XLoPIvQJLdJaV536/jTgc1V9Z6h88xJMjN3JVySA4GnM/DPXVWdWFWrut9NxzO6rc6gJQBLsAgWsXcmuSbJVYxOWy2GS+xOBR4AXNBd1nra0IEAkjwnyVbgScB5Sb4yRI5uIn3uNimbgc9U1bVDZJkvySeBfwcekWRrklcMHOnJwIuBp3b/HV3R/cU7tEOAi7qfuUsZzREsiss1FxtvMSFJjfOIQJIaZxFIUuMsAklqnEUgSY2zCCSpcRaBJDXOIpCkxv0f9YDpZzyUaKwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "binary_additive_gp_df['phenotype'].plot.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit additive parameters in binary map\n",
    "We will write a fitting function.\n",
    "We will use the general curve-fitting function [scipy.optimize.minimize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) because eventually (even if not in this homework), we'd like to extend to fitting non-linear genotype-phenotype maps.\n",
    "\n",
    "For our specific binary additive phenotype-genotype map above, we will be fitting five parameters:\n",
    "\n",
    "  - the phenotype of the wildtype\n",
    "  \n",
    "  - the effect of the mutation from `0` to `1` at each of the four sites in the sequence\n",
    "  \n",
    "So initialize a vector of parameters of the appropriate length, initially setting all to zero.\n",
    "Each entry in the vector represents one of the aforementioned parameters (in order)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0., 0., 0., 0., 0.])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "params0 = scipy.zeros(1 + binary_additive_model.seqlength)\n",
    "params0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we define a function that returns the squared difference between the actual and predicted phenotypes.\n",
    "Note that this function works **only for the specific example above**, where there is a binary genotype-phenotype map and the wildtype sequence is all `00...00`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def func_additive_binary(params):\n",
    "    \"\"\"Fit of `params` to `binary_additive_gp`.\"\"\"\n",
    "    diff = 0\n",
    "    for seq, pheno in binary_additive_gp.items():\n",
    "        predict_pheno = params[0]\n",
    "        for r, a in enumerate(seq):\n",
    "            if a == '1':\n",
    "                predict_pheno += params[r + 1]\n",
    "        diff += (pheno - predict_pheno)**2\n",
    "    return diff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we perfectly fit the map, the return value of this function should be zero.\n",
    "Here it is for our initial guess:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "79.64066968804582"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_additive_binary(params0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's fit the optimal parameters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 3.62434536 -2.22351283 -2.0563435  -3.14593724  0.73081526]\n"
     ]
    }
   ],
   "source": [
    "fit = scipy.optimize.minimize(func_additive_binary, params0)\n",
    "params = fit.x\n",
    "print(params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How good is this fit?\n",
    "Recall that the first element of the fit is the wildtype phenotype.\n",
    "Let's compare:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Actual wildtype phenotype: 3.6243453636632417\n",
      "Fit wildtype phenotype: 3.6243453564867836\n"
     ]
    }
   ],
   "source": [
    "print(f\"Actual wildtype phenotype: {binary_additive_model.wtpheno}\")\n",
    "print(f\"Fit wildtype phenotype: {params[0]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's look at the effects of the mutations.\n",
    "Recall that after the first element of the fit, each element is the value of the mutation from `0` to `1` at each site.\n",
    "Here are the actual values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>muteffect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.223513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.056344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-3.145937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.730815</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site wildtype mutation  muteffect\n",
       "0     0        0        1  -2.223513\n",
       "1     1        0        1  -2.056344\n",
       "2     2        0        1  -3.145937\n",
       "3     3        0        1   0.730815"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binary_additive_model.muteffects_table()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are the fit values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-2.22351283, -2.0563435 , -3.14593724,  0.73081526])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "params[1 : ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or more nicely formatted:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>muteffect</th>\n",
       "      <th>fit_muteffect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.223513</td>\n",
       "      <td>-2.223513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.056344</td>\n",
       "      <td>-2.056344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-3.145937</td>\n",
       "      <td>-3.145937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.730815</td>\n",
       "      <td>0.730815</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site wildtype mutation  muteffect  fit_muteffect\n",
       "0     0        0        1  -2.223513      -2.223513\n",
       "1     1        0        1  -2.056344      -2.056344\n",
       "2     2        0        1  -3.145937      -3.145937\n",
       "3     3        0        1   0.730815       0.730815"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binary_additive_muteffects_fit = (\n",
    "    binary_additive_model.muteffects_table()\n",
    "    .assign(fit_muteffect=params[1: ])\n",
    "    )\n",
    "binary_additive_muteffects_fit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plotting the correlation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f564ac6d9b0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAEGCAYAAACUzrmNAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAYhklEQVR4nO3df5AfdZ3n8edrICQoKCGJIIkRENYVvRzqLLcuXp0uCEi5QUXv1K07EDXuWix1t7UHuLnSPa0tJezd1Z5aqym0pPbwZI8UBa4oBFzXO38tgxsCCGjksEh0JYaocEJImPf98e3oMMxMfzOZ749kno+qb03/+Ez3m2Yyr+n+dH86VYUkSTMZGXQBkqThZ1hIkloZFpKkVoaFJKmVYSFJanXooAvohaVLl9bxxx8/6DIk6YByxx13/LSqlk217qAMi+OPP56xsbFBlyFJB5QkP5xunZehJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KSDhI7HtvFnQ/9jB2P7ZrzbR+Uz1lI0nxzw6ZtXLZhMwtGRtg9Ps6681ex+tTlc7Z9zywk6QC347FdXLZhM0/sHufRXXt4Yvc4l27YPKdnGIaFJB3gtu58nAUjT/91vmBkhK07H5+zfRgWknSAW7H4cHaPjz9t2e7xcVYsPnzO9mFYSNIBbskRC1l3/ioWLRjhyIWHsmjBCOvOX8WSIxbO2T7s4Jakg8DqU5dz+klL2brzcVYsPnxOgwIMC0k6aCw5YuGch8ReXoaSJLUyLCRJrQwLSVKrgYdFknOS3J9kS5LLp1h/YZLtSTY1n3cPok5Jms8G2sGd5BDgE8DrgK3A7UlurKrvTmp6bVVd3PcCJUnA4M8sTgO2VNUDVfUk8HngvAHXJEmaZNBhsRx4aML81mbZZOcn2ZzkuiQvmGpDSdYkGUsytn379l7UKknz1qDDohtfAI6vqlXARuDqqRpV1fqqGq2q0WXLlvW1QEk62A06LLYBE88UVjTLfqWqdlTV3qETrwJe2afaJEmNQYfF7cDJSU5IchjwNuDGiQ2SPH/C7Grg3j7WJ0liwHdDVdWeJBcDNwOHAJ+pqnuSfAgYq6obgUuSrAb2AI8AFw6sYEmap1JVg65hzo2OjtbY2Nigy5CkA0qSO6pqdKp1g74MJUk6ABgWkqRWhoUkqZVhIUlqZVhIkloZFpKkVoaFJKmVYSFJamVYSJJaGRaSpFaGhSSplWEhSWplWEiSWhkWkqRWhoUkqZVhIUlqZVhIkloNPCySnJPk/iRbklw+xfqFSa5t1n87yfH9r1KS5reBhkWSQ4BPAK8HTgHenuSUSc3eBeysqpOA/wZc0d8qJUmDPrM4DdhSVQ9U1ZPA54HzJrU5D7i6mb4OOCNJ+lijJM17gw6L5cBDE+a3NsumbFNVe4CfA0smbyjJmiRjSca2b9/eo3IlaX4adFjMmapaX1WjVTW6bNmyQZcjSQeVQYfFNuAFE+ZXNMumbJPkUOC5wI6+VCdJAgYfFrcDJyc5IclhwNuAGye1uRG4oJl+C/CVqqo+1ihJ896hg9x5Ve1JcjFwM3AI8JmquifJh4CxqroR+DTw10m2AI/QCRRJUh8NNCwAquom4KZJyz4wYfoJ4K39rkuS9GuDvgwlSToAGBaSpFaGhSSplWEhSWplWEiSWhkWkqRWhoUkqZVhIUlqZVhIkloZFpKkVoaFJKmVYSFJamVYSJJaGRaSpFaGhSSplWEhSWo1sLBIcnSSjUm+33xdPE27p5Jsaj6TX7kqSeqDQZ5ZXA7cVlUnA7c181N5vKpObT6r+1eeJGmvQYbFecDVzfTVwBsHWIskaQaDDItjqurHzfQ/AcdM025RkrEk30oybaAkWdO0G9u+ffucFytJ89mhvdx4kluBY6dYtXbiTFVVkppmMy+sqm1JTgS+kuSuqvrB5EZVtR5YDzA6OjrdtiRJs9DTsKiqM6dbl+QnSZ5fVT9O8nzg4Wm2sa35+kCSrwIvB54RFpKk3hnkZagbgQua6QuAGyY3SLI4ycJmeilwOvDdvlUoSQIGGxYfBV6X5PvAmc08SUaTXNW0eQkwluRO4O+Aj1aVYSFJfdbTy1AzqaodwBlTLB8D3t1MfwP4Z30uTZI0iU9wS5JadRUWSW7rZpkk6eA042WoJIuAZwFLm+E40qx6DrC8x7VJkoZEW5/Fe4F/DxwH3MGvw+IXwMd7WJckaYjMGBZV9ZfAXyb5o6r6WJ9qkiQNmW47uMeTHLV3pnn+4X09qkmSNGS6DYv3VNXP9s5U1U7gPb0pSZI0bLoNi0OS7O2vIMkhwGG9KUmSNGy6fSjvy8C1ST7VzL+3WSZJmge6DYvL6ATEHzbzG4Grpm8uSTqYdBUWVTWe5LPAV6rq/t6WJEkaNt0+wb0a2ERz6SnJqb4PW5Lmj247uD8InAb8DKCqNgEn9KooSdJw6TYsdlfVzyct8210kjRPdNvBfU+Sd9C5hfZk4BLgG70rS5I0TLo9s/gj4KXALuBzwM/pjBklSZoHZgyLJH/dTL6nqtZW1W81n/9UVU/sz46TvDXJPUnGk4zO0O6cJPcn2ZLk8v3ZpyRpdtrOLF6Z5DjgomY8qKMnfvZz33cDbwa+Nl2D5knxTwCvB04B3p7klP3cryRpH7X1WXwSuA04kacPUQ6dDu4TZ7vjqroXYMIoIlM5DdhSVQ80bT8PnAf4Hm5J6qO2M4svVNVLgM9U1YlVdcKEz6yDYh8sBx6aML8VX7okSX3XdmZxHfBK4Ddms/EktwLHTrFqbVXdMJttzrCvNcAagJUrV87lpiVp3msLi5Ekfwr8RpI/nryyqv7rTN9cVWfuT3HANuAFE+ZXNMum2td6YD3A6Oioz4BI0hxquwz1NuApOqFy5BSfXrsdODnJCUkOa+pxmBFJ6rO216reD1yRZHNVfWkud5zkTcDHgGXAF5Nsqqqzm7uvrqqqc6tqT5KLgZuBQ+j0ndwzl3VIktp1+wT3d5J8Gjiuql7f3L76qqr69Gx3XFXXA9dPsfxHwLkT5m8CbprtfiRJ+6/bJ7g/S+ev++Oa+e/hE9ySNG90GxZLq+pvgHGAqtpDpy9DkjQPdBsW/y/JEpqRZpP8Np3xoSRJ80C3fRZ/TOcupBcl+TqdTum39qwqSdJQ6XqIcuBfAS+mM+TH/XR/ViJJOsB1+wv/m1W1p6ruqaq7q2o38M1eFiZJGh4znlkkOZbOWEyHJ3k5vx5I8DnAs3pcmyRpSLRdhjobuJDOMBsTh/Z4FPjTHtUkSRoybU9wXw1cneT8qtrQp5okSUOm2w7ulyV56eSFVfWhOa5HkjSEug2LxyZMLwLeANw79+VIkoZRV2FRVf9l4nySv6Az/IckaR6Y7bMSz6LT6S1Jmge6OrNIchfNUB90hgpfBthfIUnzRLd9Fm+YML0H+EkzmKAkaR7ots/ih0kW03nF6aHAMUmoqu/0tDpJ0lDo9jLUh+k8nPcDfn05qoDf7U1ZkqRh0u1lqH8NvKiqnpyrHSd5K/BnwEuA06pqbJp2D9J5YvwpYE9Vjc5VDZKk7nQbFncDRwEPz+G+7wbeDHyqi7avraqfzuG+JUn7oNuw+Ajwj0nuBnbtXVhVq2e746q6FyBJW1NJ0oB1GxZXA1cAd9G8WrWPCrglSQGfqqr1UzVKsgZYA7By5co+lidJB79uw+KXVfXf93XjSW4Fjp1i1dqquqHLzby6qrYleR6wMcl9VfW1yY2aEFkPMDo6WpPXS5Jmr9uw+N9JPkLn1aoTL0PNeOtsVZ25H7Xt3ca25uvDSa4HTgOeERaSpN7pNixe3nz97QnLen7rbJJnAyNV9WgzfRY+OS5JfdftQ3mvnWl9kguad190LcmbgI/RGTrki0k2VdXZSY4Drqqqc4FjgOubTvBDgc9V1Zf3ZT+SpP2Xqv2/vJ/kO1X1ijmoZ06Mjo7W2NiUj21IkqaR5I7pnmWb7aizz9jHHG1HkjSE5iosvPtIkg5inllIklp1FRZJTmhZ9vU5q0iSNHS6PbPYMMWy6/ZOVNXFc1OOJGkYzXjrbJLfBF4KPDfJmyeseg6wqJeFSZKGR9tzFi+m85a8o4Dfm7D8UeA9vSpKkjRcZgyLZvymG5K8qqq+2aeaJElDpu0y1KVVtQ54R5K3T15fVZf0rDJJ0tBouwx1GbCOzutUd/a+HEnSMGoLi580YzW9E3gNPk8hSfNSW1j8FXAbcCJwx4TlofPU9ok9qkuSNETaOrg/BnwsyV9V1R/2qSZJ0pDp6qE8g0KS5re5GhtKknQQMywkSa0GFhZJrkxyX5LNSa5PctQ07c5Jcn+SLUku73edkqTBnllsBF5WVauA7wHvn9wgySHAJ4DXA6cAb09ySl+rlCQNLiyq6paq2tPMfgtYMUWz04AtVfVAVT0JfB44r181SpI6hqXP4iLgS1MsXw48NGF+a7PsGZKsSTKWZGz79u09KFGzteOxXdz50M/Y8diuQZciaZbaHsrbL0luBY6dYtXaZpBCkqwF9gDX7M++qmo9sB5gdHTU17wOiRs2beOyDZtZMDLC7vFx1p2/itWnTpn3koZYT8Oiqs6caX2SC+kMgX5GVU31C34b8IIJ8yuaZToA7HhsF5dt2MwTu8d5gnEALt2wmdNPWsqSIxYOuDpJ+2KQd0OdA1wKrK6qX07T7Hbg5CQnJDkMeBtwY79q1P7ZuvNxFow8/UdswcgIW3c+PqCKJM3WIPssPg4cCWxMsinJJwGSHJfkJoCmA/xi4GbgXuBvquqeQRWsfbNi8eHsHh9/2rLd4+OsWHz4gCqSNFs9vQw1k6o6aZrlPwLOnTB/E3BTv+rS3FlyxELWnb+KSyf1WXgJSjrwDCwsND+sPnU5p5+0lK07H2fF4sMNCukAZVio55YcsdCQkA5ww/KchSRpiBkW2ic+YCfNT16GUtd8wE6avzyzUFcmPmD36K49PLF7nEs3bPYMQ5onDAt1xQfspPnNsFBXfMBOmt8MC3Vl7wN2ixaMcOTCQ1m0YMQH7KR5xA5udc0H7KT5y7DQPvEBO2l+8jKUJKmVYSFJamVYSJJaGRaSpFaGhSSp1cDuhkpyJfB7wJPAD4B3VtXPpmj3IPAo8BSwp6pG+1mnJGmwZxYbgZdV1Srge8D7Z2j72qo61aCQpMEYWFhU1S3NO7YBvgWsGFQtkqSZDUufxUXAl6ZZV8AtSe5IsqaPNUmSGj3ts0hyK3DsFKvWVtUNTZu1wB7gmmk28+qq2pbkecDGJPdV1dem2NcaYA3AypUr56R+SVJHT8Oiqs6caX2SC4E3AGdUVU2zjW3N14eTXA+cBjwjLKpqPbAeYHR0dMptSZJmZ2CXoZKcA1wKrK6qX07T5tlJjtw7DZwF3N2/KiVJMNg+i48DR9K5tLQpyScBkhyX5KamzTHA/0lyJ/APwBer6suDKVeS5q+BPWdRVSdNs/xHwLnN9APAP+9nXZKkZxqWu6EkSUPMsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUapDv4P5wks3NK1VvSXLcNO0uSPL95nNBv+uUJA32zOLKqlpVVacCfwt8YHKDJEcDHwT+BXAa8MEki/tbpiRpYGFRVb+YMPtsoKZodjawsaoeqaqdwEbgnH7UJ0n6tUMHufMkfw78O+DnwGunaLIceGjC/NZm2VTbWgOsAVi5cuXcFipJ81xPzyyS3Jrk7ik+5wFU1dqqegFwDXDx/uyrqtZX1WhVjS5btmwuypckNXp6ZlFVZ3bZ9BrgJjr9ExNtA14zYX4F8NX9LkyStE8GeTfUyRNmzwPum6LZzcBZSRY3HdtnNcskSX00yD6LjyZ5MTAO/BD4A4Ako8AfVNW7q+qRJB8Gbm++50NV9Ugvi9rx2C627nycFYsPZ8kRC3u5K0k6YKRqqpuQDmyjo6M1Nja2z993w6ZtXLZhMwtGRtg9Ps6681ex+tQp+9Ml6aCT5I6qGp1qnU9wN3Y8tovLNmzmid3jPLprD0/sHufSDZvZ8diuQZcmSQNnWDS27nycBSNPPxwLRkbYuvPxAVUkScPDsGisWHw4u8fHn7Zs9/g4KxYfPqCKJGl4GBaNJUcsZN35q1i0YIQjFx7KogUjrDt/lZ3cksSAn+AeNqtPXc7pJy31bihJmsSwmGTJEQsNCUmaxMtQkqRWhoUkqZVhIUlqZVhIkloZFpKkVoaFJKnVQTmQYJLtdEay7ZelwE/7uL99YW2zY22zY22zMyy1vbCqpnx73EEZFv2WZGy6kRoHzdpmx9pmx9pmZ5hr28vLUJKkVoaFJKmVYTE31g+6gBlY2+xY2+xY2+wMc22AfRaSpC54ZiFJamVYSJJaGRazkOTDSTYn2ZTkliTHTdPugiTfbz4X9Km2K5Pc19R3fZKjpmn3YJK7mv+GsSGr7Zwk9yfZkuTyPtX21iT3JBlPMu0tjAM6bt3WNojjdnSSjc3P+MYki6dp91RzzDYlubHHNc14HJIsTHJts/7bSY7vZT37WNuFSbZPOFbv7ldtrarKzz5+gOdMmL4E+OQUbY4GHmi+Lm6mF/ehtrOAQ5vpK4Arpmn3ILC0z8ettTbgEOAHwInAYcCdwCl9qO0lwIuBrwKjM7QbxHFrrW2Ax20dcHkzffkMP2+P9elYtR4H4H17/80CbwOuHaLaLgQ+3s+fr24/nlnMQlX9YsLss4Gp7hI4G9hYVY9U1U5gI3BOH2q7par2NLPfAlb0ep/d6rK204AtVfVAVT0JfB44rw+13VtV9/d6P7PRZW0DOW7NPq5upq8G3tiHfc6km+MwsebrgDOSZEhqG1qGxSwl+fMkDwG/D3xgiibLgYcmzG9tlvXTRcCXpllXwC1J7kiypo817TVdbcNw3GYy6OM2nUEdt2Oq6sfN9D8Bx0zTblGSsSTfStLLQOnmOPyqTfPHy8+BJT2saV9qAzi/uVR7XZIX9KGurvha1WkkuRU4dopVa6vqhqpaC6xN8n7gYuCDw1Jb02YtsAe4ZprNvLqqtiV5HrAxyX1V9bUhqa0nuqmtCwM7boMyU20TZ6qqkkx3L/4Lm+N2IvCVJHdV1Q/mutaDwBeA/1lVu5K8l84Z0O8OuCbAsJhWVZ3ZZdNrgJt4ZlhsA14zYX4FnWvO+62ttiQXAm8AzqjmQugU29jWfH04yfV0TpH3+5feHNS2DZj419SKZtl+24f/pzNtYyDHrQsDOW5JfpLk+VX14yTPBx6eZht7j9sDSb4KvJzO9fu51s1x2Ntma5JDgecCO3pQyz7XVlUT67iKTp/QUPAy1CwkOXnC7HnAfVM0uxk4K8ni5g6Rs5plva7tHOBSYHVV/XKaNs9OcuTe6aa2u4ehNuB24OQkJyQ5jE4HZE/vnunWoI5blwZ13G4E9t7pdwHwjLOg5t/AwmZ6KXA68N0e1dPNcZhY81uAr0z3R1W/a2sCd6/VwL19qKs7g+5hPxA/wAY6vyQ20zltXN4sHwWumtDuImBL83lnn2rbQue66Kbms/euj+OAm5rpE+nciXEncA+dSx1DUVszfy7wPTp/efartjfRuYa8C/gJcPMQHbfW2gZ43JYAtwHfB24Fjm6W/+rfAvA7wF3NcbsLeFePa3rGcQA+ROePFIBFwP9qfh7/ATixH8eqy9o+0vxs3Qn8HfCb/aqt7eNwH5KkVl6GkiS1MiwkSa0MC0lSK8NCktTKsJAktTIspDmU5Kgk7+uy7SVJ7k1yTTMS6q3NSKP/Zh/3eXySd8yuYqk7hoU0t46iM6ppN94HvK6qfp/OE81U1alVde0+7vN4wLBQTxkW0iTNX+r3Jflsku81f/mfmeTrzXsbTkvyZ0n+ZML33N28F+GjwIuaM4Qrm3X/McntzeBw/7lZ9kk6D/l9KcllwP8Afqv5vhcleWWSv28GLLx575O9SU5qzkDuTPKdJC9q9vkvm+/9D/08Vpo/fChPmqT5pb+Fzl/799AZpuFO4F10hmB4J50n0B+rqr9ovuduOmNeAfxtVb2sWX4WnSEl3guEzvAO66rqa0kepPN+ip8meQ3wJ1X1hiQLgL8Hzquq7c1lqbOr6qIk3wY+WlXXJ1lE5w++0/Z+by+Pi+Y3BxKUpvZ/q+ougCT3ALdVVSW5i85ln01dbues5vOPzfwRwMnMPPjgi4GX0RnVFjovzflxMy7V8qq6HqCqnmjq24f/LGl2DAtparsmTI9PmB+n8+9mD0+/jLtomu0E+EhVfWof9h3gnqp61dMWNoMYSoNgn4U0Ow8CrwBI8grghGb5o8DEX+o3AxclOaJpu7x5F8ZM7geWJXlV8z0Lkry0qh6lM6z2G5vlC5M8a4p9SnPOsJBmZwNwdHOJ6mI6I4lSnfcRfL3p8L6yqm4BPgd8s7mEdR0tv9ir88rNtwBXJLmTziWv32lW/1vgkiSbgW/QeSnRZuCpptPbDm71hB3ckqRWnllIkloZFpKkVoaFJKmVYSFJamVYSJJaGRaSpFaGhSSp1f8HttYL+5bAKnEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "binary_additive_muteffects_fit.plot.scatter('muteffect', 'fit_muteffect')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Homework question 1\n",
    "Above we fit a binary map with a wildtype sequence of all `00.00`.\n",
    "But we could make a more general map, say with nucleotide characters and some other wildtype sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "additive_model = AdditiveModel(['A', 'C', 'G', 'T'], 'ATGC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It would again have some mutational effects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>muteffect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>-2.223513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-2.056344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-3.145937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.730815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-5.603077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>2.489624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-2.522414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.361922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-1.498741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.924216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-5.120281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-1.644834</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site wildtype mutation  muteffect\n",
       "0      0        A        C  -2.223513\n",
       "1      0        A        G  -2.056344\n",
       "2      0        A        T  -3.145937\n",
       "3      1        T        A   0.730815\n",
       "4      1        T        C  -5.603077\n",
       "5      1        T        G   2.489624\n",
       "6      2        G        A  -2.522414\n",
       "7      2        G        C  -0.361922\n",
       "8      2        G        T  -1.498741\n",
       "9      3        C        A   1.924216\n",
       "10     3        C        G  -5.120281\n",
       "11     3        C        T  -1.644834"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "additive_model.muteffects_table()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And some wildtype phenotype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.6243453636632417"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "additive_model.wtpheno"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And phenotypes for each genotype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few genotypes and their phenotypes:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAA</td>\n",
       "      <td>3.756963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAC</td>\n",
       "      <td>1.832747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAG</td>\n",
       "      <td>-3.287535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAT</td>\n",
       "      <td>0.187912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACA</td>\n",
       "      <td>5.917455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  genotype  phenotype\n",
       "0     AAAA   3.756963\n",
       "1     AAAC   1.832747\n",
       "2     AAAG  -3.287535\n",
       "3     AAAT   0.187912\n",
       "4     AACA   5.917455"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "additive_gp_df = (\n",
    "    pd.Series(additive_model.all_seqs_phenos())\n",
    "    .rename_axis('genotype')\n",
    "    .reset_index(name='phenotype')\n",
    "    )\n",
    "print('Here are the first few genotypes and their phenotypes:')\n",
    "additive_gp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And if you generalize the fitting function above, you can determine the mutational effects in this additive model.\n",
    "\n",
    "But that is **not** the homework.\n",
    "Instead I have generated a genotype-phenotype map just like the one shown above (with the same wildtype sequence of `ATGC`), but **with a different random number seed.**\n",
    "This map is in [hw_gp_list.csv](hw_gp_list.csv), which we read in here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few genotypes and their phenotypes:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAA</td>\n",
       "      <td>-5.160712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAC</td>\n",
       "      <td>0.172790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAG</td>\n",
       "      <td>-1.056697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAT</td>\n",
       "      <td>-0.308053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACA</td>\n",
       "      <td>-3.280548</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  genotype  phenotype\n",
       "0     AAAA  -5.160712\n",
       "1     AAAC   0.172790\n",
       "2     AAAG  -1.056697\n",
       "3     AAAT  -0.308053\n",
       "4     AACA  -3.280548"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hw_additive_gp_df = pd.read_csv('hw_gp_list.csv')\n",
    "print('Here are the first few genotypes and their phenotypes:')\n",
    "hw_additive_gp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Your homework question is to generalize the fitting function above to determine the mutational effects for this genotype-phenotype map.\n",
    "Turn them in as a dataframe giving the effect of each mutation at each site in the format of the similar data frames above.\n",
    "Also turn in the code you use to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
